extensions/amp-analytics/0.1/linker.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {Services} from '../../../src/services';
+import {base64UrlEncodeFromString} from '../../../src/utils/base64';
+import {crc32} from './crc32';
+
+/** @const {string} */
+const DELIMITER = '~';
+
+/**
+ * Creates a new query string parameter to be used for cross-domain ID syncing.
+ * &<paramName>=<version>~<checksum>~<key1>~<value1>~<key2>~<value2>...
+ */
+export class Linker {
+  /** @param {!../../../src/service/ampdoc-impl.AmpDoc} doc */
+  constructor(doc) {
+    this.urlReplacementService_ = Services.urlReplacementsForDoc(doc);
+  }
+
+
+  /**
+   * Creates the linker param from the given config, and returns the url with
+   * the given param attached.
+   * @param {string} version
+   * @param {Object<string,string>} pairs
+   * @return {Promise<string>}
+   */
+  create(version, pairs) {
+    return this.resolveMacros_(pairs).then(expandedPairs => {
+      return this.generateParam_(version, expandedPairs);
+    });
+  }
+
+
+  /**
+   * Go through key value pairs and resolve any macros that may exist.
+   * @param {Object<string,string>} pairs
+   * @return {Promise<Object<string, string>>}
+   */
+  resolveMacros_(pairs) {
```

Comment:
i actually think this code might be a dup of where we handle extraUrlParams.
also, this code didn't handle `${var}` style vars.

we can remove it from this class.

Reply:
I have questions about this. I found `VariableService.expandTemplate` but want to make sure we are on the same page.  We can chat Monday.

Reply:
Done.